File: swift-bridge/Sources/SmixSDK/SimRuntime.swift
// SmixSimRuntime protocol + Mock impl.
//
// SmixSDK is XCTest-free (mirror SmixRunnerCore pattern). Test author
// provides a concrete runtime impl:
//   - An XCUITest-backed runtime (real, lives in the consumer's UITest
//     target — drop-in via this protocol; wraps XCUIApplication.snapshot
//     + EventSynthesizer IOHID dance)
//   - MockSimRuntime (here): deterministic in-memory tree + event
//     log. Used by SmixSDKTests for unit-level wire verification.
//
// This protocol + Mock impl never touch the app process; they live in
// test targets only.

import Foundation
#if canImport(CoreGraphics)
import CoreGraphics
#endif

/// Mock-recorded `launchFresh(...)` call — assertable from tests.
public struct LaunchFreshCall: Sendable, Equatable {
    public let bundleId: String
    public let clearState: Bool
    public let clearKeychain: Bool
    public let appPath: String?
    public init(
        bundleId: String,
        clearState: Bool,
        clearKeychain: Bool,
        appPath: String?
    ) {
        self.bundleId = bundleId
        self.clearState = clearState
        self.clearKeychain = clearKeychain
        self.appPath = appPath
    }
}

/// Low-level simulator runtime — captures + injects events on the
/// running app.
public protocol SmixSimRuntime: Sendable {
    /// Launch the target app on the simulator.
    func launch(bundleId: String) async throws
    /// Terminate the target app on the simulator.
    func terminate(bundleId: String) async throws
    /// Capture the current a11y snapshot of the focused / front-most
    /// app window. Returns the root node (children walk applies).
    func snapshotTree() async throws -> A11yNode
    /// Synthesize a single tap at logical-points coordinate (matches
    /// `A11yNode.bounds` coordinate space).
    func synthesizeTap(at point: CGPoint) async throws
    /// Send text via the runner's keyboard fast path (Apple
    /// `_XCT_sendString:maximumFrequency:`).
    func sendString(_ text: String) async throws
    /// Press a hardware-like key.
    func pressKey(_ key: KeyName) async throws
    /// Swipe in a cardinal direction across the visible viewport
    /// (maestro navigation convention: DOWN = content scrolls down,
    /// fingers move up).
    func swipe(_ direction: SwipeDirection) async throws
    /// Capture a PNG screenshot of the visible screen.
    func screenshot() async throws -> Data
    /// Collect system popup windows (permission prompts / system
    /// alerts that live outside the app's own window).
    func systemPopups() async throws -> [A11yNode]
    /// Open a URL via `UIApplication.openURL` (deeplink, etc.).
    func openUrl(_ url: URL) async throws
    /// Launch fresh with state clearing options (mirror Rust smix-sdk
    /// `App::launch_fresh`):
    ///   - `clearState`: wipe sandbox dirs (Documents / Library) before relaunch
    ///   - `clearKeychain`: wipe keychain items for this bundle id
    ///   - `appPath`: optional override (when reinstalling from a path)
    func launchFresh(
        bundleId: String,
        clearState: Bool,
        clearKeychain: Bool,
        appPath: String?
    ) async throws
    /// Install and launch an app from an absolute `.app` path. Mirror
    /// `xcrun simctl install + launch` flow for `--app-path` workflows.
    func launchFromPath(_ appPath: String) async throws
    /// Synthesize a tap at a normalized coordinate (0..1 over the
    /// device's logical viewport). Mirrors Rust smix-sdk `tap_at_coord`.
    func synthesizeTapAtNormalized(nx: Double, ny: Double) async throws
}

/// In-memory mock runtime — used by SmixSDKTests for deterministic
/// wire verification without booting a sim.
public final class MockSimRuntime: SmixSimRuntime, @unchecked Sendable {
    /// What `snapshotTree()` returns. Default = empty container only.
    /// Mutable so tests can stage state changes between calls (e.g.
    /// snapshot 1 = empty → snapshot 2 = with target node, then
    /// Locator.toBeVisible polls and finds it on iteration 2).
    public var snapshotResult: A11yNode

    /// Log of `synthesizeTap(at:)` calls — assertable from test.
    public private(set) var tapCalls: [CGPoint] = []
    /// Log of `launch(bundleId:)` calls.
    public private(set) var launchCalls: [String] = []
    /// Log of `terminate(bundleId:)` calls.
    public private(set) var terminateCalls: [String] = []
    /// Log of `sendString(_:)` calls.
    public private(set) var sendStringCalls: [String] = []
    /// Log of `pressKey(_:)` calls.
    public private(set) var pressKeyCalls: [KeyName] = []
    /// Log of `swipe(_:)` calls.
    public private(set) var swipeCalls: [SwipeDirection] = []
    /// Counter — how many times `screenshot()` was invoked.
    public private(set) var screenshotCalls: Int = 0
    /// What `screenshot()` returns. Settable by tests (default empty).
    public var screenshotResult: Data = Data()
    /// What `systemPopups()` returns. Settable by tests (default empty).
    public var systemPopupsResult: [A11yNode] = []
    /// Log of `openUrl(_:)` calls.
    public private(set) var openUrlCalls: [URL] = []
    /// Log of `launchFresh(...)` calls.
    public private(set) var launchFreshCalls: [LaunchFreshCall] = []
    /// Log of `launchFromPath(_:)` calls.
    public private(set) var launchFromPathCalls: [String] = []
    /// Log of `synthesizeTapAtNormalized(nx:ny:)` calls as `(nx, ny)` pairs.
    public private(set) var tapAtNormalizedCalls: [CGPoint] = []

    /// Optional hook fired AFTER each snapshotTree() — lets tests
    /// stage tree transitions ("first call empty, second call has
    /// target node"). Receives the call number (0-indexed); should
    /// mutate snapshotResult for the next call.
    public var afterSnapshot: ((_ callIndex: Int) -> Void)?

    /// Internal counter for afterSnapshot hook.
    private var snapshotCallCount = 0

    public init(snapshotResult: A11yNode? = nil) {
        self.snapshotResult = snapshotResult ?? Self.emptyDefault()
    }

    public func launch(bundleId: String) async throws {
        launchCalls.append(bundleId)
    }

    public func terminate(bundleId: String) async throws {
        terminateCalls.append(bundleId)
    }

    public func snapshotTree() async throws -> A11yNode {
        let result = snapshotResult
        let idx = snapshotCallCount
        snapshotCallCount += 1
        afterSnapshot?(idx)
        return result
    }

    public func synthesizeTap(at point: CGPoint) async throws {
        tapCalls.append(point)
    }

    public func sendString(_ text: String) async throws {
        sendStringCalls.append(text)
    }

    public func pressKey(_ key: KeyName) async throws {
        pressKeyCalls.append(key)
    }

    public func swipe(_ direction: SwipeDirection) async throws {
        swipeCalls.append(direction)
    }

    public func screenshot() async throws -> Data {
        screenshotCalls += 1
        return screenshotResult
    }

    public func systemPopups() async throws -> [A11yNode] {
        systemPopupsResult
    }

    public func openUrl(_ url: URL) async throws {
        openUrlCalls.append(url)
    }

    public func launchFresh(
        bundleId: String,
        clearState: Bool,
        clearKeychain: Bool,
        appPath: String?
    ) async throws {
        launchFreshCalls.append(
            LaunchFreshCall(
                bundleId: bundleId,
                clearState: clearState,
                clearKeychain: clearKeychain,
                appPath: appPath
            )
        )
    }

    public func launchFromPath(_ appPath: String) async throws {
        launchFromPathCalls.append(appPath)
    }

    public func synthesizeTapAtNormalized(nx: Double, ny: Double) async throws {
        tapAtNormalizedCalls.append(CGPoint(x: nx, y: ny))
    }

    /// Empty container default for tests that don't care about tree
    /// content (e.g. testing the throw-on-no-match path).
    public static func emptyDefault() -> A11yNode {
        A11yNode(
            rawType: "other",
            bounds: Rect(x: 0, y: 0, w: 393, h: 852),
            enabled: true,
            visible: true,
            children: []
        )
    }
}

File: android-runner/sdk/src/main/kotlin/dev/smix/sdk/SimRuntime.kt
// SmixSimRuntime interface + MockSimRuntime impl. Kept at full API
// parity with the Swift SDK's runtime protocol.

package dev.smix.sdk

/**
 * Low-level emulator runtime — captures + injects events on the
 * running app.
 */
interface SmixSimRuntime {
    suspend fun launch(bundleId: String)
    suspend fun terminate(bundleId: String)
    suspend fun snapshotTree(): A11yNode
    suspend fun synthesizeTap(x: Double, y: Double)
    suspend fun sendString(text: String)
    suspend fun pressKey(key: KeyName)

    /** Cardinal swipe on the app's main view. */
    suspend fun swipe(direction: SwipeDirection)

    /** PNG bytes of the current sim screen. */
    suspend fun screenshot(): ByteArray

    /** Snapshot of system-level windows (alert / sheet / overlay). */
    suspend fun systemPopups(): List<A11yNode>

    /** Open a deep-link URL via the system handler. */
    suspend fun openUrl(url: String)

    /**
     * Launch fresh with state clearing options (mirrors Swift
     * `SmixSimRuntime.launchFresh`):
     *   - [clearState]: wipe sandbox dirs (Documents / Library) before relaunch
     *   - [clearKeychain]: wipe keychain items for this bundle id
     *   - [appPath]: optional override (when reinstalling from a path)
     */
    suspend fun launchFresh(
        bundleId: String,
        clearState: Boolean,
        clearKeychain: Boolean,
        appPath: String?,
    )

    /** Install and launch an app from an absolute APK / .app path. */
    suspend fun launchFromPath(appPath: String)

    /**
     * Synthesize a tap at a normalized coordinate (0..1 over the device's
     * logical viewport). Mirror Swift App.tapAtCoord escape hatch.
     */
    suspend fun synthesizeTapAtNormalized(nx: Double, ny: Double)
}

/** Mock-recorded tap call. */
data class TapCall(val x: Double, val y: Double)

/** Mock-recorded launchFresh call. */
data class LaunchFreshCall(
    val bundleId: String,
    val clearState: Boolean,
    val clearKeychain: Boolean,
    val appPath: String?,
)

/**
 * In-memory mock runtime — used by JVM unit tests for deterministic
 * wire verification without booting an emulator.
 */
class MockSimRuntime(
    var snapshotResult: A11yNode = emptyDefault(),
) : SmixSimRuntime {
    val tapCalls = mutableListOf<TapCall>()
    val launchCalls = mutableListOf<String>()
    val terminateCalls = mutableListOf<String>()
    val sendStringCalls = mutableListOf<String>()
    val pressKeyCalls = mutableListOf<KeyName>()
    val swipeCalls = mutableListOf<SwipeDirection>()
    var screenshotCalls: Int = 0
        private set
    var screenshotResult: ByteArray = ByteArray(0)
    var systemPopupsResult: List<A11yNode> = emptyList()
    val openUrlCalls = mutableListOf<String>()
    val launchFreshCalls = mutableListOf<LaunchFreshCall>()
    val launchFromPathCalls = mutableListOf<String>()
    val tapAtNormalizedCalls = mutableListOf<Pair<Double, Double>>()

    var afterSnapshot: ((Int) -> Unit)? = null
    private var snapshotCallCount = 0

    override suspend fun launch(bundleId: String) {
        launchCalls.add(bundleId)
    }

    override suspend fun terminate(bundleId: String) {
        terminateCalls.add(bundleId)
    }

    override suspend fun snapshotTree(): A11yNode {
        val idx = snapshotCallCount++
        afterSnapshot?.invoke(idx)
        return snapshotResult
    }

    override suspend fun synthesizeTap(x: Double, y: Double) {
        tapCalls.add(TapCall(x, y))
    }

    override suspend fun sendString(text: String) {
        sendStringCalls.add(text)
    }

    override suspend fun pressKey(key: KeyName) {
        pressKeyCalls.add(key)
    }

    override suspend fun swipe(direction: SwipeDirection) {
        swipeCalls.add(direction)
    }

    override suspend fun screenshot(): ByteArray {
        screenshotCalls++
        return screenshotResult
    }

    override suspend fun systemPopups(): List<A11yNode> = systemPopupsResult

    override suspend fun openUrl(url: String) {
        openUrlCalls.add(url)
    }

    override suspend fun launchFresh(
        bundleId: String,
        clearState: Boolean,
        clearKeychain: Boolean,
        appPath: String?,
    ) {
        launchFreshCalls.add(LaunchFreshCall(bundleId, clearState, clearKeychain, appPath))
    }

    override suspend fun launchFromPath(appPath: String) {
        launchFromPathCalls.add(appPath)
    }

    override suspend fun synthesizeTapAtNormalized(nx: Double, ny: Double) {
        tapAtNormalizedCalls.add(nx to ny)
    }

    companion object {
        fun emptyDefault(): A11yNode = A11yNode(
            rawType = "other",
            bounds = Rect(0.0, 0.0, 393.0, 852.0),
            enabled = true,
            visible = true,
            children = emptyList(),
        )
    }
}
